File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.go
func queryResults(limit int, queries [][]int) []int {
    n := len(queries)
    ans := make([]int, n)
    ballColorMap, colorFreq := make(map[int]int), make(map[int]int)

    for i := 0; i < n; i++ {
        ball, color := queries[i][0], queries[i][1]

        if prevColor, ok := ballColorMap[ball]; ok {
            colorFreq[prevColor]--;
            if (colorFreq[prevColor] == 0) {
                delete (colorFreq, prevColor)
            }
        }

        ballColorMap[ball] = color
        colorFreq[color]++

        ans[i] = len(colorFreq)
    }


    return ans
}

File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.cpp
class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        int n = queries.size();
        vector<int> ans(n);
        unordered_map<int, int> ballColorMap, colorFreq;

        for (int i = 0; i < n; i++) {
            int ball = queries[i][0], color = queries[i][1];

            if (ballColorMap.find(ball) != ballColorMap.end()) {
                int prevColor = ballColorMap[ball];
                colorFreq[prevColor]--;
                if (colorFreq[prevColor] == 0) colorFreq.erase(prevColor);
            }

            ballColorMap[ball] = color;
            colorFreq[color]++; 

            ans[i] = colorFreq.size();
        }

        return ans;
    }
};